### DISCUSSION TITLE
Engine / Connection seems to disconnect after some idle time

### DISCUSSION THREAD

**Discussion by bkolb249** [sqlalchemy/sqlalchemy#6632] [Usage Questions] [answered]
I have a plotly-dash application running in a docker container (based on python3.7-slim). The app is accessing a postgres database via SQLalchemy and visualizes the queried data.

After some Idle time (around 24-48 hours) the app freezes and the logs show some errors related to the database.

I cannot fully access the logs, but they contain this error:

`AttributeError: 'Connection' object has no attribute '_Connection_connection'`

and in the following, all the pieces of code which tried to query data from the database are stated (but not what exactly went wrong).

The problem was always solved with a restart of the app (and thus a new connection to the database)

I am accessing the database as follows:

```
from sqlalchemy import create_engine

...

engine = create_engine(f"postgresql+psycopg2://{user}:{passw}@{url}:{port}/{db_name}")
engine = engine.connect()

...

# using engine to send queries to database mostly like this

table1 = pd.read_sql_table("table1", con=engine)
table2 = pd.read_sql([some query], engine)

```

How can I overcome this issue? 
How could ich check whether the connection was invalidated and if so, reconnect? 

(I already tried `pool_pre_ping=True` but it had no effect)

Thank you very much in advanced

**Comment by CaselIT** [comment, marked as answer]
The issue is here:
```py
engine = create_engine(f"postgresql+psycopg2://{user}:{passw}@{url}:{port}/{db_name}")
engine = engine.connect()
```
You are creating a connection and keeping it checked oput. This prevents sqlalchemy from using any of the reconnect logic, since the connection is never returned to the pool.

I would suggest using context manager instead to execute the queries:
```py
engine = create_engine(f"postgresql+psycopg2://{user}:{passw}@{url}:{port}/{db_name}")
...
with engine.connect() as conn
  table1 = pd.read_sql_table("table1", con=conn)
  table2 = pd.read_sql([some query], conn)
```

Also since you have issues after long idle times, I would suggest first trying setting [`pool_recycle`](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine#sqlalchemy.create_engine.params.pool_recycle) before trying `pool_pre_ping`.

Documentation on this: 
- https://docs.sqlalchemy.org/en/14/core/connections.html
- https://docs.sqlalchemy.org/en/14/core/pooling.html

**Reply by bkolb249** [reply]
Context manager makes total sense, thanks. I will change the corresponding queries.

But how would `pool_recycle` help? In the documentation it says that:
> this setting causes the pool to recycle connections after the given number of seconds has passed.

What does *recycle* in that context mean? 
I would assume that this means that idle connections are returned into the pool. Which difference would that make?

**Reply by CaselIT** [reply]
> I would assume that this means that idle connections are returned into the pool.

Using a context manager the connections are returned to the pool after the context manager block.
_Recycle_ means that if a connection was opened by more than the specified number of seconds, before returning the connection to the caller, the pool will be close it and open a new one that would then be returned. This usually prevents using connections that were idle for a long time.
